torch/csrc/autograd/engine.h

```diff
@@ -112,6 +112,14 @@ struct GraphTask: std::enable_shared_from_this<GraphTask> {
 
   std::unordered_set<c10::Stream> leaf_streams;
 
+  // Per-device current and default streams of the execute() that called this GraphTask.
+  // These will be synced with leaf_streams in exec_post_processing.
+  std::vector<c10::optional<c10::Stream>> caller_current_streams_;
+  std::vector<c10::optional<c10::Stream>> caller_default_streams_;
+
+  // Collects caller_current_streams_ and caller_default_streams_
+  void stash_current_streams();
```

Comment:
cc @pritamdamania  should we add a call to this in the distributed engine to make sure it benefits from this proper synchronization as well?

Reply:
Yes, I think this makes sense.